relay-cardinality/src/cardinality.lua

```diff
@@ -0,0 +1,174 @@
+    local i = 0
+    if max_items then
+        n = math.min(n, max_items + offset)
+    end
+
+    return function()
+        local from = i * batch_size + 1 + offset
+        i = i + 1
+        if from <= n then
+            local to = math.min(from + batch_size - 1, n)
+
+            return from, to
+        end
+    end
+end
+
+
+local ACCEPTED = true
+local REJECTED = false
+local HASHES_OFFSET = 2 -- arg1: max_cardinality, arg2: expiry
+
+local working_set = KEYS[1]
+local max_cardinality = tonumber(ARGV[1])
+local expire = tonumber(ARGV[2])
+local num_hashes = #ARGV - HASHES_OFFSET
+
+local results = {
+    0, -- total cardinality
+}
+
+-- Adds elements from the table `t`, starting with index `offset` and a maximum of `max` elements
+-- to all sets in `KEYS`.
+--
+-- Returns the total amount of values added to the 'working set'.
+local function sadd(t, offset, max)
+    local added = 0;
+
+    for i = 1, #KEYS do
+        local is_working_set = i == 1
+
+        for from, to in batches(#t, 7000, offset, max) do
+            local r = redis.call('SADD', KEYS[i], unpack(t, from, to))
+            if is_working_set then
+                added = added + r
+            end
+        end
+    end
+
+    return added
+end
+
+-- Bumps to expiry of all sets by the passed expiry
+local function bump_expiry()
+    for _, key in ipairs(KEYS) do
+        redis.call('EXPIRE', key, expire)
+    end
+end
+
+
+local current_cardinality = redis.call('SCARD', working_set) or 0
+local budget = math.max(0, max_cardinality - current_cardinality)
+
+
+-- Fast Path: we have enough budget to fit all elements
+if budget >= num_hashes then
+    local added = sadd(ARGV, HASHES_OFFSET)
+    -- New current cardinality is current + amount of keys that have been added to the set
+    current_cardinality = current_cardinality + added
+
+    for _ = 1, num_hashes do
+        table.insert(results, ACCEPTED)
+    end
+
+    if added > 0 then
+        bump_expiry()
+    end
+
+    results[1] = current_cardinality
+    return results
+end
+
+-- We do have some budget
+local offset = HASHES_OFFSET
+local needs_expiry_bumped = false
+while budget > 0 and offset < #ARGV do
+    local len = math.min(#ARGV - offset, budget)
+    local added = sadd(ARGV, offset, len)
+
+    current_cardinality = current_cardinality + added
+    needs_expiry_bumped = needs_expiry_bumped or added > 0
+
+    for _ = 1, len do
+        table.insert(results, ACCEPTED)
+    end
+
+    offset = offset + len
+    budget = budget - added
+end
+
+-- Update cardinality, at this point cardinality cannot change anymore,
+-- the maximum amount of keys that can be added, have been added to the set.
+results[1] = current_cardinality
+
+-- If we ran out of budget, check the remaining items for membership
+if budget <= 0 and offset < #ARGV then
+    for arg_i = offset + 1, #ARGV do
+        local value = ARGV[arg_i]
+
+        if redis.call('SISMEMBER', working_set, value) == 1 then
```

Comment:
Can we leave a code comment here that we should use `SMISMEMBER` as soon as it's available on our cluster?